compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -516,6 +516,27 @@ impl<'a> Parser<'a> {
             token::BinOp(token::And) | token::AndAnd => {
                 make_it!(this, attrs, |this, _| this.parse_borrow_expr(lo))
             }
+            token::BinOp(token::Plus) => {
+                let mut err = this.struct_span_err(lo, "leading `+` is not supported");
+                err.span_label(lo, "unexpected `+`");
+
+                // a block on the LHS might have been intended to be an expression instead
+                let sp = this.sess.source_map().start_point(lo);
+                if let Some(sp) = this.sess.ambiguous_block_expr_parse.borrow().get(&sp) {
+                    this.sess.expr_parentheses_needed(&mut err, *sp);
+                } else {
+                    err.span_suggestion(
+                        lo,
+                        "try removing the `+`",
+                        "".to_string(),
+                        Applicability::MachineApplicable,
+                    );
```

Comment:
Let's use `span_suggestion_verbose`, so it appears on its own subdiagnostic (the new patch-style makes removal a bit easier to see, IMO), but this is a minor thing. Alternatively we could also use `tool_only_span_suggestion` here, because the removal of the `+` is kind of obvious, and all we might want is the tooltip in an editor, the label in the output doesn't _add_ that much.

Reply:
If we change the creation of `err` to be a call to `unexpected()`, then we can check here for the `look_ahead` and see if it is a literal or a binding identifier, and if so emit the suggestion and `this.bump()`.

I don't know if we can also check if the next element would be a parenthesized expression (checking for `OpenDelim(Paren)`?) and have some confidence that there won't be other parse errors.

Reply:
> Let's use `span_suggestion_verbose`, so it appears on its own subdiagnostic (the new patch-style makes removal a bit easier to see, IMO), but this is a minor thing. Alternatively we could also use `tool_only_span_suggestion` here, because the removal of the `+` is kind of obvious, and all we might want is the tooltip in an editor, the label in the output doesn't _add_ that much.

`span_suggestion_verbose` makes the fix more obvious, changed.

> I don't know if we can also check if the next element would be a parenthesized expression (checking for `OpenDelim(Paren)`?) and have some confidence that there won't be other parse errors.

I think there could be other parse errors if the next token is a parenthesis. For instance, the following expression has a parse error between the parentheses: `+(+x)`.